crypto/s2n_fips_rules.c

```diff
@@ -0,0 +1,122 @@
+    *valid = false;
+    for (size_t i = 0; i < s2n_array_len(fips_hash_algs); i++) {
+        if (fips_hash_algs[i] == hash_alg) {
+            *valid = true;
+            return S2N_RESULT_OK;
+        }
+    }
+    return S2N_RESULT_OK;
+}
+
+/* https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-52r2.pdf */
+const uint8_t fips_cipher_suite_ianas[][2] = {
+    /* 3.3.1.1.1 Cipher Suites for ECDSA Certificates */
+    { TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 },
+    { TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384 },
+    { TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256 },
+    { TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA384 },
+    { TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA },
+    { TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA },
+
+    /* 3.3.1.1.2 Cipher Suites for RSA Certificates */
+    { TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 },
+    { TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 },
+    { TLS_DHE_RSA_WITH_AES_128_GCM_SHA256 },
+    { TLS_DHE_RSA_WITH_AES_256_GCM_SHA384 },
+    { TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256 },
+    { TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA384 },
+    { TLS_DHE_RSA_WITH_AES_128_CBC_SHA256 },
+    { TLS_DHE_RSA_WITH_AES_256_CBC_SHA256 },
+    { TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA },
+    { TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA },
+    { TLS_DHE_RSA_WITH_AES_128_CBC_SHA },
+    { TLS_DHE_RSA_WITH_AES_256_CBC_SHA },
+
+    /* 3.3.1.1.2 3.3.1.2 Cipher Suites for TLS 1.3 */
+    { TLS_AES_128_GCM_SHA256 },
+    { TLS_AES_256_GCM_SHA384 },
+};
+
+S2N_RESULT s2n_fips_validate_cipher_suite(const struct s2n_cipher_suite *cipher_suite, bool *valid)
+{
+    RESULT_ENSURE_REF(cipher_suite);
+    RESULT_ENSURE_REF(valid);
+
+    *valid = false;
+    for (size_t i = 0; i < s2n_array_len(fips_cipher_suite_ianas); i++) {
+        if (fips_cipher_suite_ianas[i][0] != cipher_suite->iana_value[0]) {
+            continue;
+        }
+        if (fips_cipher_suite_ianas[i][1] != cipher_suite->iana_value[1]) {
+            continue;
+        }
+        *valid = true;
+        return S2N_RESULT_OK;
+    }
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_fips_validate_signature_scheme(const struct s2n_signature_scheme *sig_alg, bool *valid)
+{
+    RESULT_ENSURE_REF(sig_alg);
+    RESULT_GUARD(s2n_fips_validate_hash_algorithm(sig_alg->hash_alg, valid));
+    return S2N_RESULT_OK;
+}
+
+/* https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-56Ar3.pdf */
+const struct s2n_ecc_named_curve *fips_curves[] = {
+    &s2n_ecc_curve_secp256r1,
+    &s2n_ecc_curve_secp384r1,
+    &s2n_ecc_curve_secp521r1,
+};
+S2N_RESULT s2n_fips_validate_curve(const struct s2n_ecc_named_curve *curve, bool *valid)
+{
+    RESULT_ENSURE_REF(curve);
+    RESULT_ENSURE_REF(valid);
+    *valid = false;
+    for (size_t i = 0; i < s2n_array_len(fips_curves); i++) {
+        if (fips_curves[i] == curve) {
+            *valid = true;
+            return S2N_RESULT_OK;
+        }
+    }
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_fips_validate_version(uint8_t version, bool *valid)
```

Comment:
I think 1.0 and 1.1 are only allowed if they are needed for compatibility reasons
> Servers that support government-only applications8 shall be configured to use TLS 1.2 and
should be configured to use TLS 1.3 as well. These servers should not be configured to use TLS
1.1 and shall not use TLS 1.0, SSL 3.0, or SSL 2.0.

NIST.SP.800-52r2 - 3.1

Reply:
That whole section (3.1) basically boils down to "use TLS1.2 or TLS1.3, unless you have to, then use TLS1.1 or TLS1.0". But "unless you have to" is always the use case for including 1.0 or 1.1 in a policy. They're only negotiated if nothing else is possible.

Because exceptions are allowed to be made for 1.0 and 1.1, and those exceptions don't violate any other rules, I don't think we can just blanket ban them. Individual FIPS policies can just choose not to include them.

Reply:
Just another view point, I feel like our interpretation of FIPS as allowing 1.0 and 1.1 isn't really in line with other public AWS FIPS Guidance: https://aws.amazon.com/compliance/fips/
> all AWS FIPS endpoints have been updated to only accept a minimum of Transport Layer Security (TLS) 1.2 connections. This ensures that our customers who run regulated workloads can meet FedRAMP compliance requirements that mandate a minimum of TLS 1.2 encryption for data in transit.
In a way, it feels like (for AWS) this decision has already been made, and that decisions is that FIPS doesn't allow 1.0 and 1.1, even if the standard might have some legacy carve-outs for it.

Reply:
My understanding is that TLS 1.0 and TLS 1.1 are only allowed in cases where there is some justification for it.. If you have no need for TLS 1.0 but have it enabled anyways, that is not FIPS compliant.

Interpreting the broadest definition of FIPS means that our FIPS rule would really be “maybe FIPS”. Interpreting the strictest set of FIPS rules means that our FIPS rule would be “definitely FIPS”. I would prefer for our customers to be using a "definitely FIPS" rule, which would remove the ambiguity.